Update to reflect latest codec2 master

tmiw opened the pull request:
This PR contains updates based on the latest codec2 master:

1. Use ms-arm-flex6k branch from drowe67/codec2.
2. Enable 700E mode in waveform.

tmiw committed:
Updates based on latest codec2/master:

1. Use ms-arm-flex6k branch from drowe67/codec2.
2. Enable 700E mode in waveform.

tmiw commented:
I brought over the #ifdef changes from the previously used Codec2 tree. CMakeLists.txt in codec2 master can already cross compile w/o issues, so there wasn't a need to bring over the codebook changes.

tmiw committed:
Additional FDV mode updates:

1. Default to 700D instead of 1600.
2. Turn on BPF and clipping for 700D/E.
3. Set filters for all FDV modes to 2K width.

tmiw committed:
Move waveform registration/deregistration code to FreeDV API interface.

tmiw committed:
Decrease timeout to sr/packet_size.

tmiw committed:
Send meters on mode change.

tmiw commented:
@mcdermj, quick question. How exactly can one look at wherever the output() calls in the code are printing to? I don't see anything immediately obvious in SmartSDR to allow that output to go out over the network nor at least download log files that could be getting appended.

tmiw committed:
Add 6dB gain to transmitted audio to allow full use of available TX power.

tmiw committed:
Remove explicit interleave enable to match current freedv-gui logic.

tmiw committed:
Increase priority of waveform threads to improve responsiveness in SmartSDR.

tmiw committed:
Resolve warnings from previous commits.

tmiw committed:
Add needed include folder to allow codec2 to use CMSIS DSP library functions.

tmiw committed:
Suppress CMSIS build/link when compiling x86 version of waveform.

tmiw committed:
Suppress EINTR errors that aren't actually errors.

tmiw committed:
Wait for waveforms to be fully removed before waveform shutdown.

tmiw committed:
Fix build errors on ARM side.

tmiw committed:
Enable reliable_text to allow others to spot Flex users.

tmiw committed:
Fix bug causing overwriting of callsign.

tmiw committed:
Additional optimizations for the waveform.

tmiw committed:
Disable RT prioritization to avoid audio drops.

tmiw committed:
Set reliable_text ptr to NULL after deallocation.

tmiw committed:
Reset reliable_text on proper decode to prevent overflows.

tmiw committed:
Fix bug calculating semaphore wait time.

tmiw committed:
Improve receive performance.

tmiw committed:
Code cleanup.

tmiw committed:
Temporary workaround to prevent SmartSDR crashes when pushing Ctrl-C to quit workflow.

tmiw committed:
Use condition variables instead of semaphores.

tmiw committed:
Add counter to fractional timestamp field of VITA message.

tmiw committed:
Updating compile flags again.

tmiw committed:
Remove profiling from gcc options.

tmiw committed:
Enable SIMD usage in libsoxr.

tmiw committed:
Ignore patch if already applied.

amcdermond commented:
Thanks so much for all of your work @tmiw .  I really appreciate it.  I apologize, but I've been on vacation for 10 days in New England, so I haven't had a chance to review any of these commits.  I'll try to get to things in the next couple of days as I get my head a little bit more above water.

tmiw commented:
No worries! Take all the time needed.

Actually, I wonder if [recvmmsg()](https://linux.die.net/man/2/recvmmsg) works on the current Flex environment. That would allow recv/poll to be combined into a single syscall and reduce the number of times the waveform has to call into the kernel.

Also, I noticed that the older waveform code seems to change the IP for the VITA connection to localhost instead of the discovered IP if it's running on the radio itself. I didn't think that mattered but maybe it did at one point?

amcdermond commented:
The kernel is 2.6.37 and it's glibc 2.12.1.  According to https://linux.die.net/man/2/recvmmsg that should *barely* be supported.  That is not to say that I actually tried it.

tmiw commented:
On that note, sendmmsg() would be nice too but it looks like that went in for kernel 3.0.

Anyway, I'll give recvmmsg() a shot and see how that goes.

amcdermond commented:
I have some work into porting 5.4 to the flex6k, but that work is a long way to fruition right now, and may never happen.

tmiw commented:
FYI I just tried using recvmmsg() and got the following linker warning:

```
/home/mooneer/x-tools/flex6k/lib/gcc/arm-linux-gnueabihf/8.3.0/../../../../arm-linux-gnueabihf/bin/ld: src/libwaveform.a(vita-io.c.o): in function `vita_processing_loop':
/home/mooneer/smartsdr-codec2/src/vita-io.c:139: warning: warning: recvmmsg is not implemented and will always fail
[100%] Built target freedv
```

I didn't try to install the waveform but I'm presuming it'll fail as promised.

EDIT: I took a look at my ct-ng options and I'll need to build a new glibc. Will do that and report back.

amcdermond commented:
I just tried a quick test.  I checked out this code which purports to test recvmmsg: https://github.com/pasqualirb/recvmsg_vs_recvmmsg

It compiled with the warning you mentioned:
```
anna.mcdermond@flextop recvmsg_vs_recvmmsg % make
/Volumes/toolchain/flex6k/bin/flex6k-gcc -Wall -D_GNU_SOURCE   -c -o main.o main.c
/Volumes/toolchain/flex6k/bin/flex6k-gcc -Wall -D_GNU_SOURCE   -c -o receive.o receive.c
/Volumes/toolchain/flex6k/bin/flex6k-gcc -lrt  main.o receive.o   -o main
/Volumes/toolchain/flex6k/bin/../lib/gcc/arm-linux-gnueabi/10.2.0/../../../../arm-linux-gnueabi/bin/ld: receive.o: in function `do_receive_recvmmsg':
receive.c:(.text+0x1a): warning: warning: recvmmsg is not implemented and will always fail
```

I uploaded the test app to a flex6k and attempted to run it:
```
root@flex6k:~# ./main -b 128 -c 100 4444
Using recvmmsg()
Elapsed time: 537.326550 ms
```

During the run I performed a `tcpdump -i lo 'port 4444'` to ensure that we were actually getting packets sent and received.
```
22:02:45.680449 IP localhost.localdomain.4444 > localhost.localdomain.4444: UDP, length 13
22:02:45.680475 IP localhost.localdomain.4444 > localhost.localdomain.4444: UDP, length 13
22:02:45.680504 IP localhost.localdomain.4444 > localhost.localdomain.4444: UDP, length 13
22:02:45.680531 IP localhost.localdomain.4444 > localhost.localdomain.4444: UDP, length 13
22:02:45.680558 IP localhost.localdomain.4444 > localhost.localdomain.4444: UDP, length 13
```
 I would expect to get errors if it always fails.

tmiw committed:
Use new branch for Codec2.

tmiw committed:
Updates:

1. Use recvmmsg() for VITA receive thread.
2. Reenable profiling on ARM target.
3. Use localhost if waveform is running on radio.

tmiw committed:
Add freedv_tx2 and freedv_rx2 targets to enable testing of codec2 optimizations without SmartSDR.

tmiw commented:
Last night I was able to rebuild the toolchain with glibc 2.17 and minimum kernel 2.6.33 and got rid of the warning that way. I'm still seeing the issue described in https://github.com/drowe67/codec2/pull/242, though, but maybe a bit less than before? I might try switching the send side out for sendmsg() with a multi-element msg_iov array and seeing how that goes.

amcdermond commented:
I would be careful with the glibc versions.  The actual version on the box is:

`GNU C Library (EGLIBC) stable release version 2.12.2, by Roland McGrath et al.`

We have run into issues trying to push that too far.  If you'd like, we do have a Docker container with the "official" toolchain we actually use to build firmware loads that I can probably share.  If you'd prefer, I can probably also share a ct-ng config file with some notes.

tmiw commented:
Yeah, I figured much further than 2.17 would cause it to try to use different syscalls than actually intended or something. So far the waveform seems to be doing okay with that version so I'll leave it for now. 👍

tmiw committed:
Fix indentation for iovec code.

tmiw committed:
Use aio_write() to asynchronously write to the VITA socket.

drowe67 commented on src/freedv-processor.c:
```diff
@@ -253,52 +279,89 @@ static void *_sched_waveform_thread(void *arg)
+                for (int radio_samples = freedv_nin(params->fdv);
+                     ringbuf_bytes_used(params->process_buffer) >= radio_samples * sizeof(short);
+                     radio_samples = freedv_nin(params->fdv)) {
 
```
This `for()` construct concerns me, would rather see `nin` sampled just after `freedv_rx()`, rather than on the fly like this as the point at when it is sampled is not obvious (e.g. beginning or end of loop).  Suggest we keep it simple, and just like the examples in `codec2/demo` as a first step.

tmiw replied:
I tweaked how that was done since I was already adding debug related preprocessor stuff.

drowe67 commented on src/freedv-processor.c:
```diff
@@ -310,68 +373,78 @@ static void *_sched_waveform_thread(void *arg)
                                   mod_out, tx_modem_samples, NULL,
                                   resample_buffer, tx_modem_samples * SAMPLE_RATE_RATIO, &odone);
 
+                    // Make samples louder by 6dB to compensate for lower than expected power output otherwise on Flex (e.g. setting the power slider to max only gives 30-40W out on the 6300 using 700D).
```
maybe some asserts on `odone` to make sure the values are sensible?

tmiw replied:
Added. Per soxr docs odone should always be <= tx_modem_samples.

drowe67 commented on src/freedv-processor.c:
```diff
@@ -310,68 +373,78 @@ static void *_sched_waveform_thread(void *arg)
+                        resample_buffer[index] *= tx_scale_factor;
+                    }
+
                     ringbuf_memcpy_into (tx_output_buffer, resample_buffer, odone * sizeof(float));
```
I'd avoid boosting the level for a first pass, until the signal integrity issues are sorted out.

It would be good to check the float values are in the range +/-1 as required by the API.

tmiw replied:
I wrapped that around an #ifdef and kept [the associated #define](https://github.com/mcdermj/smartsdr-codec2/pull/17/commits/d276daea2eb38ca3c0caa741eb7bb4c9b98aa6ad#diff-69912da7d14c9a86ccbdd1b20f8d4b2e76322047462e0c6535869ee58bca6dc7R260) commented out for the time being.

drowe67 commented:
A couple of test modes might be useful:
1. Analog pass through, just take input speech, don't call `freedv_tx`, but do all the downsampling and upsampling.  That will test out the entire framework with FreeDV out of the loop.  Same for the rx side. This would be a good general purpose demo for the API too.
1. Sine wave Tx mode, along the lines of what I suggested [here](https://github.com/drowe67/codec2/pull/242#issuecomment-948952606)

tmiw committed:
Add various debugging options for the waveform.

tmiw commented:
> A couple of test modes might be useful:
> 
> 1. Analog pass through, just take input speech, don't call `freedv_tx`, but do all the downsampling and upsampling.  That will test out the entire framework with FreeDV out of the loop.  Same for the rx side. This would be a good general purpose demo for the API too.
> 2. Sine wave Tx mode, along the lines of what I suggested [here](https://github.com/drowe67/codec2/pull/242#issuecomment-948952606)

Added sine wave and analog passthrough for both RX and TX. 👍

tmiw committed:
Reduce inbound packet buffer since SmartSDR isn't going to send that many at once.

tmiw committed:
Add 8K TX file output.

tmiw committed:
Added some timing output to validate required execution time.

tmiw committed:
Use DIGU/DIGL for underlying_mode.

tmiw committed:
Buffer and do 3 TX calls at a time.

tmiw committed:
Buffer a few more TX packets before sending them.

tmiw committed:
Cap gained values to [-1, 1].

tmiw committed:
Bump up version to 2.0.2.

tmiw committed:
Massive cleanup inside freedv_processor.c:

1. Reduce the number of ringbufs required.
2. Move common processing code into separate functions.
3. Reduce required heap space.

tmiw committed:
Bump up version to 2.0.3.

tmiw committed:
Tweak ringbuf sizes.

tmiw committed:
WIP code to support the Teensy based FreeDV dongle.

tmiw committed:
Latest dongle protocol.

tmiw committed:
Update dongle_protocol again to reflect removal of explicit RX/TX mode change message.

tmiw committed:
Lower amplification to 3dB instead of 5dB due to clipping issues.

tmiw committed:
Various changes:

1. Support callsign encoding if using the Teensy.
2. Reinitialize the Teensy when switching between TX and RX.

tmiw committed:
Add missed sending of reliable_text config.

tmiw committed:
Update dongle_protocol.

tmiw committed:
update dongle_protocol

tmiw commented:
@amcdermond, hey there! I'm finally updated to SmartSDR 3.3.29 (unfortunately the initial upgrade failed and I needed a replacement SD card so there was a bit of a delay). A few things:

1. On the 6300 I have here, the code as is in this PR doesn't seem to run on the radio at all (when installed via File->Waveforms, probably the same with the Windows installer as well), dongle enabled or no. It seems to run on my local machine as well as on the radios of some other people who've been running this, however, so I'm not sure what's up.
2. When it's running, I've noticed an issue where waveforms (well, this one anyway) seems to only be able to use LSB/DIGL. I've tried disabling FDVL in the code to no avail as well as using "USB" instead of "DIGU" for the definition of FDVU.

Anyway, based on the release notes for 3.3.29 it doesn't seem like it would have caused any of the above issues but if you have any insight that would definitely be appreciated. 👍

tmiw committed:
Increase TX output for SmartSDR 3.4.

tmiw committed:
WIP checkin.

tmiw committed:
Remove soxr.

tmiw committed:
Use codec2 FIFO for 8K samples as well as adjust receive timing on VITA thread.

tmiw committed:
Bump up number of send packets that can be queued.
